### PR TITLE
Partition-based hash join

### PR DESCRIPTION
**Pull request by don-tay**
## Changes in this PR

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change of an existing feature
- [ ] Refactor/Performance enhancements

## Overview

- Partition-based hash join. Implementation partitions both table A and B into `(B-1)` partitions. It then uses a page-based nested loop to join the tables in the respective partitions.

### CODE REVIEW

**Comment by don-tay** [comment]
Test queries:
```
select sid,eid from student,enroll where studentid=sid
```

**Comment by don-tay** [comment]
Bug in query
```
select sid,sname from student, enroll, section where sid=studentid and sectionid=sectid
```

**Comment by don-tay** [comment]
stack trace of bug above
image

**Comment by don-tay** [comment]
Test queries
2 table join
```
select sid,sname,dname from student,dept where majorid=did
```
4 table join
```
select sid,sname,dname,title,grade from student,dept,course,enroll where sid=studentid and deptid=did and majorid=did
```
5 table join
```
select sid,sname,dname,title,prof,grade from student,dept,course,section,enroll where sid=studentid and sectionid=sectid and courseid=cid and deptid=did and majorid=did
```

File: src/simpledb/materialize/HashJoinScan.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    * Create a hashjoin scan for the two underlying hash buckets.
+    * 
+    * @param b1       the LHS hashed buckets
+    * @param b2       the RHS hashed buckets
+    * @param fldname1 the LHS join field
+    * @param fldname2 the RHS join field
+    * @param tx       the current transaction
+    */
+   public HashJoinScan(Transaction tx, List<TempTable> b1, List<TempTable> b2, String fldname1, String fldname2) {
+      this.tx = tx;
```
Comment (xinyee20) [inline src/simpledb/materialize/HashJoinScan.java:31]: Transaction tx wasn't used in the functions so maybe can remove it from HashJoinScan constructor too

**Review by xinyee20 — APPROVED** [review]
lgtm

File: src/simpledb/materialize/HashJoinScan.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import simpledb.query.*;
+import simpledb.tx.Transaction;
+
+/**
+ * The Scan class for the <i>hashjoin</i> operator.
+ * 
+ * @author Edward Sciore
+ */
+public class HashJoinScan implements Scan {
+   private Transaction tx;
```
[the remark this answers is not in the document]
Comment (don-tay) [reply src/simpledb/materialize/HashJoinScan.java]: Will do, thank you!

**Pull request merged**